chatbot-core/data/preprocessing/utils/filter_functions.py

```diff
@@ -0,0 +1,140 @@
+from bs4 import BeautifulSoup, Comment
+
+def extract_page_content_container(soup, class_name):
+    """
+    Extracts the HTML content of the first div with the given class name.
+
+    Parameters:
+    - soup (BeautifulSoup): Parsed HTML document.
+    - class_name (str): Class name of the div to extract.
+
+    Returns:
+    - str: HTML string of the matching div, or empty string if not found.
+    """
+    content_div = soup.find("div", class_=class_name)
+    if content_div:
+        return str(content_div)
+    return ""
+
+
+def remove_container_by_class(content, class_name):
+    """
+    Given HTML content and a class name, removes the content inside the div
+    with that class name and returns the remaining HTML.
+
+    Parameters:
+    - content (str): HTML string.
+    - class_name (str): Class name of the div to extract.
+
+    Returns:
+    - str: HTML filtered from the div with class class_name.
+    """
+    soup = BeautifulSoup(content, "html.parser")
+    content_div = soup.find("div", class_=class_name)
+
+    if content_div:
+        content_div.decompose() 
+
+    return str(soup)
+
+def remove_tags(content, tags_to_remove=None):
+    """
+    Removes all specified tags from the given HTML content.
+    
+    Parameters:
+    - content (str): The HTML string to clean.
+    - tags_to_remove (list of str): Tag names to remove (e.g., ["img", "script"]).
+    
+    Returns:
+    - str: Cleaned HTML content with specified tags removed.
+    """
+    if tags_to_remove is None:
+        tags_to_remove = ["img", "script", "style", "iframe", "object", "embed", "form"]  # Default tags to remove
+
+    soup = BeautifulSoup(content, "html.parser")
+
+    for tag in tags_to_remove:
+        for element in soup.find_all(tag):
+            element.decompose()
+
+    return str(soup)
+
+def remove_edge_navigation_blocks(content):
+    """
+    Removes the first navigation block and the last one (plus any following elements)
+    from the direct children of `.col-lg-9`, based on expected structure.
+
+    Parameters:
+    - content (str): HTML string.
+
+    Returns:
+    - str: HTML filtered from any not wanted navigation elements.
+    """
+    soup = BeautifulSoup(content, "html.parser")
+    container = soup.find("div", class_="col-lg-9")
+    if not container:
+        return str(soup)
+
+    def is_navigation_block(div):
+        if div.name != "div":
+            return False
+        children = div.find_all(recursive=False)
+        return (
+            len(children) == 1 and
+            children[0].name == "div" and
+            "row" in children[0].get("class", []) and
+            "body" not in children[0].get("class", [])
+        )
+
+    # Remove top navigation block
+    children = container.find_all(recursive=False)
```

Comment:
This can lead to skipped elements or unexpected behavior, so, what we do is we iterate over a copy. ``` for div_to_remove in children[i:]:
    div_to_remove.extract()
```